src/linker/Linker/TypeReferenceExtensions.cs

```diff
@@ -12,46 +13,94 @@ public static string GetDisplayName (this TypeReference type)
 			if (type == null)
 				return string.Empty;
 
-			var builder = new System.Text.StringBuilder ();
-			if (type.DeclaringType != null)
-				builder.Append (type.DeclaringType.GetDisplayName ()).Append (".");
-
-			if (type.HasGenericParameters) {
-				builder.Append (type.Name.Substring (0, type.Name.IndexOf ('`')));
-				builder.Append ('<');
-				for (int i = 0; i < type.GenericParameters.Count - 1; i++)
-					builder.Append ($"{type.GenericParameters[i]},");
-
-				builder.Append ($"{type.GenericParameters[type.GenericParameters.Count - 1]}>");
-			} else if (type is ArrayType arrayType) {
-				void parseArrayDimensions (ArrayType at)
-				{
-					builder.Append ('[');
-					for (int i = 0; i < at.Dimensions.Count - 1; i++)
-						builder.Append (',');
-
-					builder.Append (']');
+			var sb = new StringBuilder ();
+			Stack<TypeReference> genericArguments = null;
+			while (true) {
+				switch (type) {
+				case ArrayType arrayType:
+					ParseArrayType (arrayType, sb);
+					break;
+				case GenericInstanceType genericInstanceType:
+					int arguments = int.Parse (genericInstanceType.Name.Substring (genericInstanceType.Name.IndexOf ('`') + 1));
+					genericArguments = new Stack<TypeReference> (genericInstanceType.GenericArguments);
+					ParseGenericInstanceType (genericArguments, arguments, sb);
+					sb.Insert (0, genericInstanceType.Name.Substring (0, genericInstanceType.Name.IndexOf ('`')));
+					break;
+				default:
+					if (type.HasGenericParameters) {
+						if (!int.TryParse (type.Name.Substring (type.Name.IndexOf ('`') + 1), out int genericParameters)) {
+							sb.Insert (0, type.Name);
+							break;
+						}
+
+						ParseGenericParameters (type.GenericParameters.TakeLast (genericParameters).ToList (), genericArguments, genericParameters, sb);
+						sb.Insert (0, type.Name.Substring (0, type.Name.IndexOf ('`')));
+						break;
+					}
+
+					sb.Insert (0, type.Name);
```

Comment:
Same here - why prepend... this can easily be done by appending, right?
(The reason is that Insert is inherently slower as it has to move the existing content first).